cmd/internal/loginprovider/tryinorderloginprovider.go

```diff
@@ -26,8 +26,13 @@ func NewTryInOrderLoginProvider(logger loggo.Logger, providers ...api.LoginProvi
 }
 
 type tryInOrderLoginProviders struct {
-	providers []api.LoginProvider
-	logger    loggo.Logger
+	providers  []api.LoginProvider
+	logger     loggo.Logger
+	loginToken string
+}
+
+func (p *tryInOrderLoginProviders) Token() string {
```

Comment:
Maybe the same godoc here.

Reply:
I am genuinely worried this is not sufficient for long running sessions or the possibility of re-use. In an OIDC world tokens are refreshed at a regular cadence. We aren't really handling that fact. We are also taking context away from the provider. Part of the call to the provider in the context that is has available is the time at which the information is requested.

If I was going to solve this I would probably store this as a closure that defer the call into the provider.

I am also noticing a theme that `Token` probably needs the opportunity to return an error to the call site. As two of the login providers have a complex state that needs to be setup and the call to `Token` can't be controlled. i.e what happens if this is called and the closure is nil? Does empty string mean that a token does or doesn't exist?

Reply:
I'm not sure I follow all your concerns here. Could you provide an example?

From I understand, you mean that the token returned could expire by the time the client uses it elsewhere? That is certainly possible if there was some client that performed an initial connection and then some hours later tried to make a call that required basic auth.

That to me speaks more to the problematic nature of the API and how one endpoint /api has a different authentication method versus /log. If they were the same there wouldn't be a need for `Token` as `Login` would always handle things properly.

As to your solution, I understand what you mean with the closure, the different providers can determine whether the token is still valid at the time the token is requested. The password for example would always be valid, the session token on the other hand should be able to be inspected to see if it is valid and if not restart the login flow.

But as it stands, I don't know that we this issue with long running sessions currently and we might never. Happy to refactor to something that caters to sessions if you think we should head in that direction.

Reply:
Yep, I understand that the likelihood of this being an issue given how short our login contexts are from the client. I still don't want to rely on that fact. Would rather get the design correct. If we can let the actual provider handle the call via closures it gives us full context inside the provider.

It also means the aggregate doesn't ever need to take on special case logic. Given my proposal above for introducing an error type that will be highly context aware.

Reply:
Realised better what you meant after our discussion and added a closure around the underlying provider's token method. Let me know if this is sufficient.